src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,165 @@
 
 package com.googleinterns.zoomtube.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import com.googleinterns.zoomtube.data.Lecture;
 import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
 
-/**
- * Provides information on a lecture.
- */
+/** Provides information on a lecture. */
 @WebServlet("/lecture")
 public class LectureServlet extends HttpServlet {
+  /* Used to generate a Pattern for a Video URL. */
+  private static final String YOUTUBE_VIDEO_URL_PATTERN =
+      "(?<=watch\\?v=|/videos/|embed\\/|youtu.be\\/|\\/v\\/|\\/e\\/|"
+      + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
+      + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
+
+  /* Name of input field used for lecture name in lecture selection page. */
+  private static final String NAME_INPUT = "name-input";
+  
+  /* Name of input field used for lecture video link in lecture selection page. */
+  private static final String LINK_INPUT = "link-input";
+
+  private static final String REDIRECT_URL = "/lecture-view.html";
+
+  /* Pattern used to create a matcher for a video ID. */
+  private static Pattern videoUrlGeneratedPattern;
+
+  private static DatastoreService datastore;
+
   @Override
   public void init() throws ServletException {
-    // TODO: Implement LectureServlet.
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    videoUrlGeneratedPattern = Pattern.compile(YOUTUBE_VIDEO_URL_PATTERN);
   }
 
   @Override
+  // TODO: Check if URL is valid.
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
+    String videoUrl = optionalVideoUrl.isPresent() ? optionalVideoUrl.get() : "";
+    Optional<Entity> existingEntity = checkUrlInDatabase(videoUrl);
```

Comment:
instead of passing an empty string to checkUrlInDatabase, just pass the Optional result. Then you can convert it in there.